File: src/main/java/com/github/clevernucleus/playerex/factory/DamageFactory.java
package com.github.clevernucleus.playerex.factory;

import java.util.function.BiConsumer;

import com.github.clevernucleus.dataattributes_dc.api.DataAttributesAPI;
import com.github.clevernucleus.playerex.api.ExAPI;
import com.github.clevernucleus.playerex.api.damage.DamageFunction;
import com.github.clevernucleus.playerex.api.damage.DamagePredicate;

import net.minecraft.entity.AreaEffectCloudEntity;
import net.minecraft.entity.damage.DamageTypes;
import net.minecraft.entity.effect.StatusEffects;
import net.minecraft.entity.projectile.thrown.PotionEntity;

public final class DamageFactory {
	public static final DamageFactory STORE = new DamageFactory();

	private DamageFactory() {
	}

	public void forEach(BiConsumer<DamagePredicate, DamageFunction> registry) {
		registry.accept((living, source, damage) -> source.isOf(DamageTypes.ON_FIRE),
				(living, source, damage) -> DataAttributesAPI // <- ON_FIRE? unsure about implementation, but should be
																// right
						.ifPresent(living, ExAPI.FIRE_RESISTANCE, damage, value -> (float) (damage * (1.0 - value))));
		registry.accept((living, source, damage) -> source.isOf(DamageTypes.FREEZE),
				(living, source, damage) -> DataAttributesAPI.ifPresent(living, ExAPI.FREEZE_RESISTANCE, damage,
						value -> (float) (damage * (1.0 - value))));
		registry.accept((living, source, damage) -> source.isOf(DamageTypes.LIGHTNING_BOLT),
				(living, source, damage) -> DataAttributesAPI.ifPresent(living, ExAPI.LIGHTNING_RESISTANCE, damage,
						value -> (float) (damage * (1.0 - value))));
		registry.accept(
				(living, source, damage) -> living.hasStatusEffect(StatusEffects.POISON)
						&& source.isOf(DamageTypes.MAGIC)
						&& damage <= 1.0F,
				(living, source, damage) -> DataAttributesAPI.ifPresent(living, ExAPI.FREEZE_RESISTANCE, damage,
						value -> (float) (damage * (1.0 - value))));
		registry.accept(
				(living, source, damage) -> source.isOf(DamageTypes.WITHER) || (source.isOf(DamageTypes.INDIRECT_MAGIC)
						&& (source.getSource() instanceof PotionEntity
								|| source.getSource() instanceof AreaEffectCloudEntity)),
				(living, source, damage) -> {
					return DataAttributesAPI.ifPresent(living, ExAPI.WITHER_RESISTANCE, damage, value -> {
						if (source.isOf(DamageTypes.WITHER) && living.isUndead())
							return 0.0F;
						if (source.getName().equals("indirectMagic") && source.getSource() instanceof PotionEntity
								&& living.isUndead())
							return damage;
						return (float) (damage * (1.0 - value));
					});
				});
	}
}

File: src/main/kotlin/com/bibireden/playerex/factory/DamageFactory.kt
package com.bibireden.playerex.factory

import com.bibireden.data_attributes.api.DataAttributesAPI
import com.bibireden.playerex.api.attribute.PlayerEXAttributes
import com.bibireden.playerex.api.damage.DamageFunction
import com.bibireden.playerex.api.damage.DamagePredicate
import net.minecraft.world.damagesource.DamageTypes
import net.minecraft.world.effect.MobEffects
import net.minecraft.world.entity.AreaEffectCloud
import net.minecraft.world.entity.projectile.ThrownPotion

object DamageFactory {
    fun forEach(registry: (damagePredicate: DamagePredicate, damageFunction: DamageFunction) -> Unit)
    {
        registry.invoke({ _, source, _ -> source.`is`(DamageTypes.ON_FIRE) }, { entity, _, damage ->
            DataAttributesAPI.getValue(PlayerEXAttributes.FIRE_RESISTANCE, entity).map { damage * (1 - it) }.map(Double::toFloat).orElse(damage)
        })
        registry.invoke({ _, source, _ -> source.`is`(DamageTypes.FREEZE)}, { living, _, damage ->
            DataAttributesAPI.getValue(PlayerEXAttributes.FREEZE_RESISTANCE, living).map { damage * (1 - it) }.map(Double::toFloat).orElse(damage)
        })
        registry.invoke({ _, source, _ -> source.`is`(DamageTypes.LIGHTNING_BOLT)}, { living, _, damage ->
            DataAttributesAPI.getValue(PlayerEXAttributes.LIGHTNING_RESISTANCE, living).map { damage * (1 - it) }.map(Double::toFloat).orElse(damage)
        })
        registry.invoke(
            { living, source, damage -> living.hasEffect(MobEffects.POISON) && source.`is`(DamageTypes.MAGIC) && damage <= 1.0F },
            {living, _, damage -> DataAttributesAPI.getValue(PlayerEXAttributes.POISON_RESISTANCE, living).map { damage * (1 - it) }.map(Double::toFloat).orElse(damage) }
        )
        registry.invoke({_, source, _ -> source.`is`(DamageTypes.WITHER) || (source.`is`(DamageTypes.INDIRECT_MAGIC) && ((source.directEntity is ThrownPotion) || (source.directEntity is AreaEffectCloud))) }, {
            living, source, damage ->
            DataAttributesAPI.getValue(PlayerEXAttributes.WITHER_RESISTANCE, living).map {
                if (source.`is`(DamageTypes.WITHER) && living.isInvertedHealAndHarm) return@map 0.0F
                if (source.`is`(DamageTypes.INDIRECT_MAGIC) && source.directEntity is ThrownPotion && living.isInvertedHealAndHarm) return@map damage
                (damage * 1 - it).toFloat()
            }.orElse(damage)
        })
    }
}